Fix #3327: Create app layer mechanism for saving checkpoints.

MaskedCarrot opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

Fixes #3327 

This PR is stacked on top of PR #3403

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.


## Testing the changes made in this PR

1. Checkpointing is disabled by default until a mechanism to resume checkpoints is also implemented. However it can be enabled by changing the value of the variable isCheckpointingEnabled and shouldSavePartialProgresss to true in TopicLessonsFragment.

2. Since size limit imposed on checkpoint database is set to 2MB, therefore the ProgressDatabaseFull dialog won’t be visible until checkpoint database exceeds 2MB in size So to view the ProgressDatabaseFull dialog box, enable checkpointing and then set the size limit to 0 in ExplorationStorageModule.


## Tests

Test | Espresso test result
---| ---
ExplorationActivityTest | ![Screenshot from 2021-07-06 10-23-43](https://user-images.githubusercontent.com/53645584/124545231-4a445000-de46-11eb-842d-e2cd248e0661.png)


## Dialog boxes after this PR

This PR introduces the following three dialog boxes: 

1. **StopExplorationDialog**: Visible to the user when the exit an exploration that is saved and size of checkpoint database has not exceeded.
2. **UnsavedProgressDialog**: Visible to the user when the progress is not saved ( this can happen if checkpointing fails or the lesson was previously completed). This dialog box is also visible to the user when they exit a practice session.
3. **ProgressDatabaseFullDialog**: Visible to the user when the progress is saved but the checkpoint database has exceeded the size limit.

<img src="https://user-images.githubusercontent.com/53645584/124349148-6b6d2c80-dc0b-11eb-9a69-5fd44f5a587b.jpg" align="left" height="450" width="230" >

<img src="https://user-images.githubusercontent.com/53645584/124349494-e551e580-dc0c-11eb-8fef-021998a45df9.jpg" align="left" height="450" width="230" >

<img src="https://user-images.githubusercontent.com/53645584/124349480-d2d7ac00-dc0c-11eb-9a4e-bc9f5c6357cd.jpg" align="left" height="450" width="230" >

MaskedCarrot committed:
add controller and tests

MaskedCarrot committed:
NIT

Commit:
added tests for constant

Commit:
moved enum to a different file

Commit:
fixed lint

MaskedCarrot committed:
Merge branch 'develop' into add-controller-for-checkpointing

Commit:
fixed comments

Commit:
Merge branch 'add-controller-for-checkpointing' of https://github.com/oppia/oppia-android into add-controller-for-checkpointing

Commit:
saving

Commit:
added mech to save checkpoints

Commit:
lint fix

Commit:
added module to UI-tests

Commit:
added exploration storage module to ui tests

Commit:
nit

MaskedCarrot committed:
Merge branch 'add-module-to-tests' into saving-checkpoints

Commit:
added module in missed files

Commit:
lint fix

Commit:
fixed saving

Commit:
added app layer saving mechanism

Commit:
added app layer saving mechanism

Commit:
lint fix

Commit:
fix lint in domain layer

Commit:
lint fix

Commit:
nit changes

Commit:
Merge branch 'add-module-to-tests' into saving-checkpoints

Commit:
merge base branch

Commit:
lint fix

Commit:
implemented dialog boxes

Commit:
lint fix

Commit:
added tests for saving checkpoints

Commit:
nit changes

Commit:
lint fix

Commit:
nit changes

Commit:
lint fix

Commit:
lint fix

Commit:
fixed failing test

Commit:
imporved approach and added tests

Commit:
nit changes

Commit:
nit changes

Commit:
fixed function names

Commit:
fixed failing domain tests and some nits

Commit:
completed saving approach

Commit:
added more tests and fixed failing tests

Commit:
checkpointing disabled by default

anandwana001 commented:
Assign me back once the UI layer gets reviewed, if there were any UI changes, the test needs to be reviewed again to prevent multiple reviews here.

Commit:
fix

Commit:
nit

rt4914 commented:
@aggarwalpulkit596 Can you take first pass at it? If not, feel free to assign back to me. (Not removing my assignment as I will review it tomorrow.)

Commit:
added more tests

MaskedCarrot committed:
Merge branch 'saving-checkpoints' into app-layer-mech-for-checkpointing

Commit:
added default value to boolean

Commit:
Merge remote-tracking branch 'localrepo/app-layer-mech-for-checkpointing' into app-layer-mech-for-checkpointing

BenHenning commented:
Since the base PR is still up in the air, not reviewing this until that's finalized. Please reassign to me at that point.

Commit:
improved mechanism to save checkpoints

Commit:
nit changes

Commit:
Merge branch 'saving-checkpoints-alternate' into app-layer-mech-for-checkpointing

Commit:
nit

Commit:
Merge branch 'saving-checkpoints-alternate' into app-layer-mech-for-checkpointing

Commit:
completed app layer implementation

Commit:
nit

Commit:
fixed failing test

Commit:
nit changes

Commit:
suppressed ExperimantalCoroutinesApi warning

Commit:
Merge branch 'saving-checkpoints-alternate' into app-layer-mech-for-checkpointing

Commit:
added more tests

Commit:
nit

MaskedCarrot committed:
Merge branch 'develop' into saving-checkpoints-alternate

Commit:
nit

MaskedCarrot committed:
Merge branch 'saving-checkpoints-alternate' into app-layer-mech-for-checkpointing

MaskedCarrot commented:
The base PR is in review and it should not have any major changes now. So @rt4914 @vinitamurthi @anandwana001  @BenHenning can you please this PR as well.

aggarwalpulkit596 commented on app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -195,6 +211,27 @@ class ExplorationActivityPresenter @Inject constructor(
+    if (oldestExplorationId != null) {
+      explorationDataController.deleteExplorationProgressById(
+        ProfileId.newBuilder().setInternalId(internalProfileId).build(),
+        oldestExplorationId!!
```
Why do we need to use `!!` complier should handle this since we have already added a null check

MaskedCarrot replied:
`oldestExplorationId`  can be null but `deleteExplorationProgressById` expects it to be not null. `oldestExplorationId` will be null if there was some error while retrieving the oldest checkpoint.

MaskedCarrot replied:
`!!` removing this gives an error.

MaskedCarrot replied:
Removed `!!`

aggarwalpulkit596 reviewed (commented):
LGTM. Just a nit-pick.

vinitamurthi commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationDataController.kt:
```diff
@@ -80,6 +85,38 @@ class ExplorationDataController @Inject constructor(
+    }
+  }
+
+  /** Function to fetch the details of the oldest saved exploration for a specified profileId.*/
```
Add `@param` and `@return` annotations (here and below)

MaskedCarrot replied:
done

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt:
```diff
@@ -954,114 +992,701 @@ class ExplorationActivityTest {
-    if (isOnRobolectric()) {
-      val dataSource = createAudioDataSource(
-        explorationId = RATIOS_EXPLORATION_ID_0, audioFileName = "content-en-057j51i2es.mp3"
+  // TODO(#89): Check this test case too.
```
can we update the comment in a bit more understandable way?

MaskedCarrot replied:
done

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt:
```diff
@@ -1108,9 +1752,10 @@ class ExplorationActivityTest {
       FirebaseLogUploaderModule::class, FakeOppiaClockModule::class, PracticeTabModule::class,
       DeveloperOptionsStarterModule::class, DeveloperOptionsModule::class,
-      ExplorationStorageModule::class
+      TestExplorationStorageModule::class
```
Do we need to update this in all test files? if any?

MaskedCarrot replied:
` TestExplorationStorageModule` is added to the test file where we need to change the size of the checkpoint database to test a few specific checkpointing features. 

The test files that need this module are:

1. ExplorationActivityTest
2. ExplorationDataControllerTest
3. ExplorationProgressControllerTest
4. ExplorationCheckpointControllerTest

anandwana001 replied:
Are all test in this class passing on robolectric?

MaskedCarrot replied:
Yes, all tests added here are passing on robolectric

anandwana001 reviewed (changes requested):
nit things

Commit:
nit

Commit:
nit

Commit:
nit

Commit:
nit

Commit:
nit

Commit:
removed commit

Commit:
added more tests

MaskedCarrot committed:
Merge branch 'saving-checkpoints-alternate' into app-layer-mech-for-checkpointing

Commit:
removed visibleForTesting

Commit:
nit

Commit:
Revert "removed visibleForTesting"

This reverts commit 796330039af873f6521fa3f4a9d9492a1b2fae26.

anandwana001 reviewed (changes requested):
Assign me back once approved the UI part, and then I will review the test cases.

Commit:
added TODO for the issue

Commit:
removed git add .

Commit:
Merge remote-tracking branch 'localrepo/app-layer-mech-for-checkpointing' into app-layer-mech-for-checkpointing

Commit:
nit

Commit:
Merge branch 'saving-checkpoints-alternate' into app-layer-mech-for-checkpointing

Commit:
nit

Commit:
Merge branch 'develop' into saving-checkpoints-alternate

Commit:
Merge branch 'saving-checkpoints-alternate' into app-layer-mech-for-checkpointing

Commit:
added missing module in develop

Commit:
Merge branch 'saving-checkpoints-alternate' into app-layer-mech-for-checkpointing

Commit:
nit

BenHenning commented:
@MaskedCarrot please clarify what aspect of this PR you'd like me to review since I don't have any code owners required here.

Also, please fix the failing CI checks that can be fixed (given that this is a chained PR, the base branch check is expected to fail).

MaskedCarrot commented:
@BenHenning can you please take a look at ExplorationActivityPresenter and ExplorationProgressController

Commit:
nit

MaskedCarrot commented:
@rt4914 please take a look.

MaskedCarrot committed:
Merge branch 'saving-checkpoints-alternate' into app-layer-mech-for-checkpointing

Commit:
Merge remote-tracking branch 'origin/develop' into saving-checkpoints-alternate

Commit:
reverted changes back in StateDeck.kt

Commit:
Merge branch 'saving-checkpoints-alternate' into app-layer-mech-for-checkpointing

Commit:
fixed failing tests

Commit:
fixed failing test

Commit:
fixed failing tests

Commit:
Merge branch 'saving-checkpoints-alternate' into app-layer-mech-for-checkpointing

Commit:
added more tests

MaskedCarrot committed:
Merge branch 'saving-checkpoints-alternate' into app-layer-mech-for-checkpointing

rt4914 commented on app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivity.kt:
```diff
@@ -16,18 +16,21 @@ import org.oppia.android.app.model.State
-private const val TAG_STOP_EXPLORATION_DIALOG = "STOP_EXPLORATION_DIALOG"
+const val TAG_UNSAVED_EXPLORATION_DIALOG = "UNSAVED_EXPLORATION_DIALOG"
+const val TAG_STOP_EXPLORATION_DIALOG = "STOP_EXPLORATION_DIALOG"
+const val TAG_PROGRESS_DATABASE_FULL_DIALOG = "PROGRESS_DATABASE_FULL_DIALOG"
```
These consts are getting used only in `Presenter`. Lets shift them there and make it private.

MaskedCarrot replied:
done

aggarwalpulkit596 reviewed (approved):
LGTM

rt4914 commented on app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -49,6 +54,11 @@ class ExplorationActivityPresenter @Inject constructor(
 
+  private var isCheckpointingEnabled: Boolean = false
+
+  private lateinit var oldestExplorationId: String
```
I think this was named considering the `oldest progress` concept in mind but here we can use `explorationId` as it will always be either null or Id of the current exploration.

WDYT?

Same for Title.

MaskedCarrot replied:
@rt4914 This will not be the id of the current exploration instead it is the explorationId of the exploration which has the oldest saved progress in the database.

MaskedCarrot replied:
renamed this to `oldestCheckpointExplorationId`

rt4914 replied:
Ack.

rt4914 commented on app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -195,6 +211,27 @@ class ExplorationActivityPresenter @Inject constructor(
     ) as HintsAndSolutionExplorationManagerFragment?
   }
 
+  fun deleteCurrentProgressAndStopExploration() {
```
Add `KDoc` to these public functions.

MaskedCarrot replied:
done

rt4914 commented on app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -195,6 +211,27 @@ class ExplorationActivityPresenter @Inject constructor(
+    stopExploration()
+  }
+
+  fun deleteOldestSavedProgressAndStopExploration() {
```
Add `KDoc` to these public functions.

MaskedCarrot replied:
done

rt4914 commented on app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -227,6 +264,18 @@ class ExplorationActivityPresenter @Inject constructor(
     }
   }
 
+  fun backButtonPressed() {
```
Add `KDoc` to these public functions.

MaskedCarrot replied:
done

rt4914 commented on app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -299,4 +348,122 @@ class ExplorationActivityPresenter @Inject constructor(
+          oldestExplorationTitle = it.getOrThrow().explorationTitle
+        } else if (it.isFailure()) {
+          oppiaLogger.e(
+            "Exploration Activity",
```
It should be `ExplorationActivity`

MaskedCarrot replied:
done

rt4914 commented on app/src/main/java/org/oppia/android/app/player/stopplaying/ProgressDatabaseFullDialogFragment.kt:
```diff
@@ -0,0 +1,79 @@
+import androidx.fragment.app.DialogFragment
+import org.oppia.android.R
+
+const val OLDEST_SAVED_EXPLORATION_TITLE_ARGUMENT_KEY =
```
Make this private

MaskedCarrot replied:
done

rt4914 commented on app/src/main/java/org/oppia/android/app/testing/ExplorationTestActivityPresenter.kt:
```diff
@@ -61,8 +61,9 @@ class ExplorationTestActivityPresenter @Inject constructor(
-              EXPLORATION_ID, /* backflowScreen= */
-              null
+              EXPLORATION_ID,
+              /* backflowScreen= */null,
```
Nit space
`/* backflowScreen= */ null,`

MaskedCarrot replied:
done

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt:
```diff
@@ -127,8 +135,9 @@ import javax.inject.Singleton
 )
 class ExplorationActivityTest {
-  @get:Rule
-  val accessibilityTestRule = AccessibilityTestRule()
```
Why was this removed? We need this check to make sure the app is accessible by all and tests pass on it.

MaskedCarrot replied:
done

rt4914 reviewed (changes requested):
@MaskedCarrot Please add exact details to PR description about what this PR does, what steps should reviewer do to check this, what is the output, etc.? Considering that the issue also does not contain information anyone who is not fully aware about this project won't understand what's going on.

All following details to your output:
1. Screenshot that all Espresso Test files which have been created/modified in this PR are passing.
2. Screenshot of UI if any.

Commit:
nit

MaskedCarrot commented:
@rt4914 PTAL. I have added the requested changes and also updated the PR description with instructions on how to test out the changes made in this PR.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt:
```diff
@@ -955,113 +996,700 @@ class ExplorationActivityTest {
+      "exploration/$explorationId/assets/audio/$audioFileName"
+  }
+
+  private fun waitForTheView(viewMatcher: Matcher<View>): ViewInteraction {
```
Don't use this function, we use testcoroutinedispatchers class and ignore using this wait fucntion

MaskedCarrot replied:
This function is not being used by any of the tests added in this PR. `waitForMatch` has a TODO to solve this.

Commit:
added tests and exempted ui files

Commit:
nit

Commit:
nit

rt4914 reviewed (approved):
@MaskedCarrot Just make sure that the base branch is develop.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt:
```diff
@@ -955,113 +996,700 @@ class ExplorationActivityTest {
+    addException.invoke(/* obj= */ null, dataSource, exception)
+  }
+
+  private fun scrollToViewType(viewType: StateItemViewModel.ViewType) {
```
remove this function as this is not getting used by any function

MaskedCarrot replied:
done

anandwana001 reviewed (approved):
LGTM

BenHenning commented on app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -299,4 +359,125 @@ class ExplorationActivityPresenter @Inject constructor(
+  }
+
+  /**
+   * This function listens to the result of [ExplorationDataController.getExplorationById].
```
Function KDocs shouldn't start with "This function...". Instead, start with the verb.

Ditto everywhere in this PR.

MaskedCarrot replied:
done

BenHenning commented on app/src/main/java/org/oppia/android/app/player/stopplaying/StopStatePlayingSessionWithSavedProgressListener.kt:
```diff
@@ -0,0 +1,13 @@
+  fun deleteCurrentProgressAndStopSession()
+
+  /**
+   * leave the currently playing session and overwrite the oldest saved progress with the current
```
The verb tense & capitalization is wrong here & above--please update to be consistent with the broader codebase & our conventions.

MaskedCarrot replied:
done

BenHenning commented on app/src/main/java/org/oppia/android/app/player/stopplaying/ProgressDatabaseFullDialogFragment.kt:
```diff
@@ -0,0 +1,79 @@
+  private fun createMaximumStorageCapacityReachedDialogDescription(
+    oldestSavedExplorationTitle: String
+  ): String {
+    return "Saved progress for the lesson \"$oldestSavedExplorationTitle\" will be deleted."
```
Please never create strings programmatically--we should always use variable interpolation from strings.xml so that translators can rearrange the string correctly. Ditto anywhere else this might've been done in this PR.

MaskedCarrot replied:
changed this to use variable interpolation.

BenHenning commented on app/src/main/res/values/strings.xml:
```diff
@@ -31,10 +31,18 @@
   <string name="stop_exploration_dialog_cancel_button">Cancel</string>
   <string name="stop_exploration_dialog_leave_button">Leave</string>
+  <string name="progress_database_full_dialog_title">Maximum storage capacity reached</string>
+  <string name="progress_database_full_dialog_continue_button">Continue</string>
```
This experience seems confusing to me--'Continue' doesn't indicate to the user that they are about to delete progress. What do the PRD & mocks specify in this case?

MaskedCarrot replied:
The mocks say continue. Although they also say click to continue in the dialog description. What should the buttons be here?

![Screenshot from 2021-07-15 11-27-51](https://user-images.githubusercontent.com/53645584/125736406-10d35be4-f54e-43dc-ab6d-7ea4bd7d8418.png)

MaskedCarrot replied:
And also if I add the line "Click continue to proceed", do I add it in continuation with the dialog description or after leaving a line like shown in the mock?

BenHenning replied:
I'm not sure that I understand @MaskedCarrot. What does the current UI look like?

MaskedCarrot replied:
The current UI looks like this 

<img src="https://user-images.githubusercontent.com/53645584/124349480-d2d7ac00-dc0c-11eb-9a4e-bc9f5c6357cd.jpg" align="left" height="450" width="230" >

BenHenning replied:
As discussed over chat, we resolved the confusion & the code has been cleaned up to be a lot clearer.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -105,6 +111,29 @@ class ExplorationProgressController @Inject constructor(
+      // size limit or the current exploration contains unsaved progress.
+      val currentCheckpointState = explorationProgress.checkpointState
+      if (currentCheckpointState == CheckpointState.CHECKPOINT_SAVED_DATABASE_EXCEEDED_LIMIT) {
+        throw CheckpointDatabaseOverflowException(
```
We shouldn't use exceptions in this way. Exceptions should only be used in exceptional cases. Here, we expect these errors to potentially happen which means returning an enum status would probably be clearer & avoids the potential confusion that these exceptions are catching actual errors vs. legitimate states.

Also, conceptually it's a successful state if the checkpoint can't be saved for some detected reason. AsyncResult's error state should also be exception/reserved for unexpected issues that are detected.

MaskedCarrot replied:
changed the approach to dataProvider approach so this is above changes are no longer needed.

BenHenning reviewed (changes requested):
Thanks @MaskedCarrot. Had a few concerns that I think you need to address looking at the specific files that you asked me to. I think my comment from your base PR regarding the DataProvider approach is pertinent here since it can largely simplify how the data flow mechanism is working for exiting the exploration.

If possible, we should try to avoid blocking the exit on a DataProvider since that may result in user-perceived jank.

vinitamurthi reviewed (approved):
LGTM thanks!

Commit:
changed approach to stop exploration and nit changes

Commit:
nit

Commit:
Merge branch 'saving-checkpoints-alternate' into app-layer-mech-for-checkpointing

Commit:
nit

Commit:
Merge branch 'saving-checkpoints-alternate' into app-layer-mech-for-checkpointing

Commit:
changed approach to stop exploration and nits

Commit:
lint fix

Commit:
nit

Commit:
Merge branch 'saving-checkpoints-alternate' into app-layer-mech-for-checkpointing

Commit:
nits in comments

Commit:
Merge branch 'saving-checkpoints-alternate' into app-layer-mech-for-checkpointing

Commit:
nit

Commit:
applied suggestion in stateDeck

Commit:
Merge branch 'saving-checkpoints-alternate' into app-layer-mech-for-checkpointing

Commit:
nit

MaskedCarrot commented on app/src/main/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerActivity.kt:
```diff
@@ -125,4 +123,11 @@ class QuestionPlayerActivity :
   override fun dismissConceptCard() {
     questionPlayerActivityPresenter.dismissConceptCard()
   }
+
+  override fun deleteCurrentProgressAndStopSession() {
+    questionPlayerActivityPresenter.stopTrainingSession()
+  }
+
+  // This function is not needed because the no progress is saved in a training session.
+  override fun deleteOldestProgressAndStopSession() {}
```
These are the two methods that have names that can be a bit misleading in QuestionPlayerActivity.

MaskedCarrot reviewed (commented):
@BenHenning PTAL.

Commit:
nit

MaskedCarrot reviewed (commented):
These are the methods whose names can be a bit misleading in questionPlayerActivity

MaskedCarrot commented on app/src/main/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerActivity.kt:
```diff
@@ -125,4 +123,14 @@ class QuestionPlayerActivity :
   override fun dismissConceptCard() {
     questionPlayerActivityPresenter.dismissConceptCard()
   }
+
+  override fun deleteCurrentProgressAndStopSession() {
+    // No progress is saved in training sessions.
+    questionPlayerActivityPresenter.stopTrainingSession()
+  }
+
+  override fun deleteOldestProgressAndStopSession() {
+    // This function is not needed because their is no progress that is being saved in training
+    // sessions.
+  }
```
These names can be a bit misleading.

BenHenning replied:
So I think my concern here is that the training session experience shouldn't even show the flow that results in progress being deleted (i.e. I'd expect the old dialog to show up & neither of these methods to get called).

MaskedCarrot replied:
The user still sees the old dialog box with every thing is exactly the same as it was from the user's viewpoint.
But the code for this has changed since from explorations progress also has to be deleted. 

Since all the existing dialog boxes have progress associated with them (saved, unsaved and databaseFull), so should I file a new issue that after milestone 1 the training session should have a new dialog box?

BenHenning commented on app/src/main/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerActivity.kt:
```diff
@@ -125,4 +123,14 @@ class QuestionPlayerActivity :
+  }
+
+  override fun deleteOldestProgressAndStopSession() {
+    // This function is not needed because their is no progress that is being saved in training
```
```suggestion
    // This function is not needed because there is no progress that is being saved in training
```

MaskedCarrot replied:
done

BenHenning commented on scripts/assets/test_file_exemptions.textproto:
```diff
@@ -265,7 +265,10 @@ exempted_file_path: "app/src/main/java/org/oppia/android/app/player/state/testin
 exempted_file_path: "app/src/main/java/org/oppia/android/app/player/state/testing/StateFragmentTestViewModel.kt"
 exempted_file_path: "app/src/main/java/org/oppia/android/app/player/stopplaying/RestartPlayingSessionListener.kt"
 exempted_file_path: "app/src/main/java/org/oppia/android/app/player/stopplaying/StopExplorationDialogFragment.kt"
+exempted_file_path: "app/src/main/java/org/oppia/android/app/player/stopplaying/ProgressDatabaseFullDialogFragment.kt"
+exempted_file_path: "app/src/main/java/org/oppia/android/app/player/stopplaying/UnsavedExplorationDialogFragment.kt"
```
These actually should have tests to verify that their buttons perform the expected results.

MaskedCarrot replied:
There are tests to verify the buttons works correctly in ExplorationActivityTest

BenHenning replied:
As discussed in chat, nearly all files should have tests to verify them in isolation but we're making an exception here for this PR since the dialogs are still being tested via ExplorationActivityTest.

BenHenning reviewed (commented):
Thanks @MaskedCarrot. I had a few concerns left, but otherwise the PR LGTM.

Commit:
nit

Commit:
Merge branch 'saving-checkpoints-alternate' into app-layer-mech-for-checkpointing

Commit:
nit

MaskedCarrot committed:
Merge branch 'develop' into app-layer-mech-for-checkpointing

Commit:
fixed dialog box

Commit:
Merge remote-tracking branch 'localrepo/app-layer-mech-for-checkpointing' into app-layer-mech-for-checkpointing

Commit:
disabled checkpointing, it was enabled for testing locally

BenHenning reviewed (approved):
Thanks @MaskedCarrot! LGTM for reviewed files.

BenHenning commented:
Given that past tests have passed, I'm fine with enabling auto-merge here since Gradle tests should be sufficient for the latest changes.
